### PR TITLE
fix(NODE-3386): listCollections result type definition

### PR DESCRIPTION
**Pull request by nbbeeken**
The listCollections cursor now has a generic that defaults to a flexible CollectionInfo type. Typescript can't change a return type bases on a runtime boolean so I show example usage in the added types test of how one can use the API pretty succinctly.

### CODE REVIEW

**Comment by addaleax** [comment]
> Typescript can't change a return type bases on a runtime boolean

Fwiw, it can come pretty close: [TypeScript playground](https://www.typescriptlang.org/play?#code/FASwdgLgpgTgZgQwMZQAQDkEFsoHswA2AnqgN7CqWpjZQBcqAzhDOAOYDcwAvsKJLEQpUAMQCuBAgEkwcXKigAPaGAAmjDLXzEyFKjRwNmrMJz2UIRAA70mLdl17AkBBIw0ARAEa6qqAiDMAMK4klBIECD4jAAUcCAE0DAMAErhuDCqADzG7AA0qAhgRAB8BbhWkdEMpNS0APKERAwsYmjcAJQMAAowuFiBUFmYONpEANoAuiVcfgHBoQThVWCx8YmwqemZOfamBUWl5ZVRqzV1OI3EDIgEjO1dqL39g1nikjJyUzPm-oEQITCEVOawSSS2SAy2Vy+0KxTKqAqK0Y5wMUCuzVQXlwoSgRVQnR6fQG9yyMXe0lk8gAPppRk0Ot9ZlR5gDFssQXEwZtUGlITsYWwDvDjsjUQ0mgxsbj8YSnsTXuSJJS5KhaSM8Azvr4-FQYFAIGIYGB5S97gA6fWMUIANygMXGpDRdAA5IoXQVLDYGC6AF4u7iTDrMyi8JxuIhgJCoOBiKMrGM4mIdHWUSGrCAXTU6AC8hQA7ggQJmYmAoPnUN5k+bWYClsDojFSNwCrU0RiWjA2gSOsHfunmDHleBVXmEIXi6hS+XK15q7X2Q3Vk2W2Qsx2Ywg7g8+34B5m44wjWgxxOS2WK1WOjX-nWOY3m6315LUABZBAQAAWlqKqn6ydQEpUAABnNABWHtdyoAB6aCszGcZgMmc0vSgDhUFg1AsgAWlQNgQDtDRUMYJBWEqBQYD6GBfljSQR1wRDkNQ9DMJw6h5FgKjfkww9j0YlDrDQjC4LYrdrQoqiNB4IA)

**Comment by nbbeeken** [comment]
Oo thanks for the insight, I'll give that a try

**Review by addaleax — APPROVED** [review]

**Review by emadum — APPROVED** [review]
LGTM 👍

File: test/types/list_collections.test-d.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+// We default to the CollectionInfo result type
+expectType<ListCollectionsCursor<Pick<CollectionInfo, 'name' | 'type'> | CollectionInfo>>(
+  db.listCollections()
+);
+// By default it isn't narrowed to either type
+expectNotType<ListCollectionsCursor<Pick<CollectionInfo, 'name' | 'type'>>>(db.listCollections());
+expectNotType<ListCollectionsCursor<CollectionInfo>>(db.listCollections());
+
+// Testings each argument variation
```
Comment (dariakp) [inline test/types/list_collections.test-d.ts:18]:
```suggestion
// Testing each argument variation
```

**Review by dariakp — CHANGES_REQUESTED** [review]
small typo, otherwise LGTM :)

**Comment by nbbeeken** [comment]
Rebased, should clear up the unrelated error in the last CI run

**Pull request merged**